{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "25e814ee-3dc7-4605-a5b9-32df47f2f07b",
   "metadata": {},
   "source": [
    "# **Preparation of Validation Datasets**\n",
    "\n",
    "To evaluate the performance of our multi-omics ranking methods, we compiled independent validation datasets from reputable databases across multiple biological entities, including genes, miRNAs, methylation markers, and pathways. These datasets serve as external benchmarks to assess the biological relevance of our predicted features for each disease-specific dataset: **Alzheimer's Disease**, **Breast Cancer**, and **Lung Adenocarcinoma**.\n",
    "\n",
    "**Gene Validation**\n",
    "\n",
    "* **GeneCards**: A comprehensive gene database providing integrated annotations, used to extract high-confidence disease-associated genes.\n",
    "* **Comparative Toxicogenomics Database (CTD)**: Used to validate gene–disease associations based on curated and inferred evidence.\n",
    "\n",
    "**miRNA Validation**\n",
    "\n",
    "* **HMDD (Human microRNA Disease Database)**: Employed to confirm known miRNA–disease associations based on experimental evidence.\n",
    "\n",
    "**Methylation Validation**\n",
    "\n",
    "* **EWAS Atlas**: Used to extract disease-associated epigenetic markers (CpGs or genes) identified in Epigenome-Wide Association Studies (EWAS).\n",
    "\n",
    "**Pathway Validation**\n",
    "\n",
    "* **CTD**: Provided curated information on disease-related biological pathways.\n",
    "* **G\\:Profiler**: Used for functional enrichment and pathway mapping of predicted features, enabling cross-validation of pathway-level relevance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60d393c3-f793-4a92-a5f8-a8477688f529",
   "metadata": {},
   "source": [
    "### Load Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2680d87e-6894-454d-856b-77a2eba6e13f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7db91f2-458f-4082-bf7b-ae4ee9b18767",
   "metadata": {},
   "source": [
    "### Load Raw Datasets and Prepare"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "962579ed-5af6-4efb-beea-6c15c79f4703",
   "metadata": {},
   "source": [
    "#### 1. Prepare GeneCards Valdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0bfe122c-5fa0-4b46-a99d-1fda1059d515",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APP</td>\n",
       "      <td>147.905914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PSEN1</td>\n",
       "      <td>125.910118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>APOE</td>\n",
       "      <td>99.596542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PSEN2</td>\n",
       "      <td>80.182045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAPT</td>\n",
       "      <td>60.629429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Entity       Score\n",
       "0    APP  147.905914\n",
       "1  PSEN1  125.910118\n",
       "2   APOE   99.596542\n",
       "3  PSEN2   80.182045\n",
       "4   MAPT   60.629429"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for alzheimer disease\n",
    "alzhemer_gene_cards = pd.read_csv('../downloads/val_data/GeneCards-SearchResults-Alzheimers.csv') \n",
    "alzhemer_gene_cards = alzhemer_gene_cards[['Gene Symbol', 'Relevance score']]\n",
    "alzhemer_gene_cards.rename(columns = {'Gene Symbol': 'Entity', 'Relevance score': 'Score'}, inplace=True) \n",
    "alzhemer_gene_cards.sort_values(by='Score', ascending=False)\n",
    "alzhemer_gene_cards.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "03f22829-7ba8-4a13-aad8-9751fe97a79b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATM</td>\n",
       "      <td>313.776306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PALB2</td>\n",
       "      <td>283.848755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRIP1</td>\n",
       "      <td>268.905914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>261.717804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>252.202255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Entity       Score\n",
       "0    ATM  313.776306\n",
       "1  PALB2  283.848755\n",
       "2  BRIP1  268.905914\n",
       "3  BRCA2  261.717804\n",
       "4  CHEK2  252.202255"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for Breast Cancer disease\n",
    "bc_gene_cards = pd.read_csv('../downloads/val_data/GeneCards-SearchResults-BreastCancer.csv') \n",
    "bc_gene_cards = bc_gene_cards[['Gene Symbol', 'Relevance score']]\n",
    "bc_gene_cards.rename(columns = {'Gene Symbol': 'Entity', 'Relevance score': 'Score'}, inplace=True) \n",
    "bc_gene_cards.sort_values(by='Score', ascending=False)\n",
    "bc_gene_cards.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2789cc2-2e7a-4b35-bd9e-ae2e17864fae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAPT</td>\n",
       "      <td>128.200134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNCA</td>\n",
       "      <td>18.441431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRN</td>\n",
       "      <td>17.680164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STH</td>\n",
       "      <td>16.352251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LRRK2</td>\n",
       "      <td>15.583508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Entity       Score\n",
       "0   MAPT  128.200134\n",
       "1   SNCA   18.441431\n",
       "2    GRN   17.680164\n",
       "3    STH   16.352251\n",
       "4  LRRK2   15.583508"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for Lung Cancer disease\n",
    "psp_gene_cards = pd.read_csv('../downloads/val_data/GeneCards-SearchResults-PSP.csv') \n",
    "psp_gene_cards = psp_gene_cards[['Gene Symbol', 'Relevance score']]\n",
    "psp_gene_cards.rename(columns = {'Gene Symbol': 'Entity', 'Relevance score': 'Score'}, inplace=True) \n",
    "psp_gene_cards.sort_values(by='Score', ascending=False)\n",
    "psp_gene_cards.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e02eee8a-1c7e-46e8-a06b-ed6b131cfff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# aggregate these results\n",
    "val_datasets = [\n",
    "    pd.concat([pd.DataFrame({'Disease':['AD']*alzhemer_gene_cards.shape[0]}), alzhemer_gene_cards], axis =  1),\n",
    "    pd.concat([pd.DataFrame({'Disease':['BRCA']*bc_gene_cards.shape[0]}), bc_gene_cards], axis =  1),\n",
    "    pd.concat([pd.DataFrame({'Disease':['PSP']*psp_gene_cards.shape[0]}), psp_gene_cards], axis =  1)\n",
    "]\n",
    "genecards_gene_val_data = pd.concat(val_datasets, axis = 0)  \n",
    "genecards_gene_val_data['Entity'] =genecards_gene_val_data['Entity'].str.strip()\n",
    "genecards_gene_val_data.to_csv('../data/val_data/genecards_val_data.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5dc3fcc-1867-4de5-9c29-4d6844bd3bd8",
   "metadata": {},
   "source": [
    "#### 2. Prepare CTD Valdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d2511b64-29a4-425e-a70d-fc5ad286f9c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD</td>\n",
       "      <td>Immune System</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD</td>\n",
       "      <td>Signal Transduction</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AD</td>\n",
       "      <td>Innate Immune System</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AD</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AD</td>\n",
       "      <td>Metabolism of proteins</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Disease                  Entity  Score\n",
       "0      AD           Immune System     31\n",
       "1      AD     Signal Transduction     28\n",
       "2      AD    Innate Immune System     23\n",
       "3      AD              Metabolism     23\n",
       "4      AD  Metabolism of proteins     22"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load pathway val data for Alzheimer's disease\n",
    "ad_ctd_disease_pathways = pd.read_csv('../downloads/val_data/CTD_D000544_pathways_20250721063406.csv') \n",
    "ad_ctd_disease_pathways['Score'] = ad_ctd_disease_pathways['Association inferred via'].apply(lambda x: len(x.split('|'))) \n",
    "ad_ctd_disease_pathways['Entity'] = ad_ctd_disease_pathways.Pathway \n",
    "ad_ctd_disease_pathways = ad_ctd_disease_pathways[['Entity', 'Score']] \n",
    "ad_ctd_disease_pathways.sort_values('Score',  ascending = False, inplace = True) \n",
    "ad_ctd_disease_pathways.insert(0, 'Disease', ['AD']*ad_ctd_disease_pathways.shape[0])\n",
    "ad_ctd_disease_pathways['Entity'] = ad_ctd_disease_pathways['Entity'].str.strip()\n",
    "ad_ctd_disease_pathways.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dac94dc2-d463-4169-b66e-e145000d68ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>Signal Transduction</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>Immune System</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>Pathways in cancer</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Disease               Entity  Score\n",
       "0    BRCA  Signal Transduction    150\n",
       "1    BRCA        Immune System    109\n",
       "2    BRCA      Gene Expression    100\n",
       "3    BRCA           Metabolism     94\n",
       "4    BRCA   Pathways in cancer     78"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load pathway val data\n",
    "brc_ctd_disease_pathways = pd.read_csv('../downloads/val_data/CTD_D001941_pathways_20250904053420.csv') \n",
    "brc_ctd_disease_pathways['Score'] = brc_ctd_disease_pathways['Association inferred via'].apply(lambda x: len(x.split('|'))) \n",
    "brc_ctd_disease_pathways['Entity'] = brc_ctd_disease_pathways.Pathway \n",
    "brc_ctd_disease_pathways = brc_ctd_disease_pathways[['Entity', 'Score']] \n",
    "brc_ctd_disease_pathways.sort_values('Score',  ascending = False, inplace = True) \n",
    "brc_ctd_disease_pathways.insert(0, 'Disease', ['BRCA']*brc_ctd_disease_pathways.shape[0])\n",
    "brc_ctd_disease_pathways['Entity'] = brc_ctd_disease_pathways['Entity'].str.strip()\n",
    "brc_ctd_disease_pathways.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e56fde5c-268a-4a7c-a4a9-1868e6e96012",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSP</td>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PSP</td>\n",
       "      <td>Herpes simplex infection</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PSP</td>\n",
       "      <td>mRNA Splicing</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PSP</td>\n",
       "      <td>mRNA Splicing - Major Pathway</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PSP</td>\n",
       "      <td>Processing of Capped Intron-Containing Pre-mRNA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Disease                                           Entity  Score\n",
       "0     PSP                              Alzheimer's disease      2\n",
       "2     PSP                         Herpes simplex infection      2\n",
       "3     PSP                                    mRNA Splicing      2\n",
       "4     PSP                    mRNA Splicing - Major Pathway      2\n",
       "5     PSP  Processing of Capped Intron-Containing Pre-mRNA      2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load pathway val data\n",
    "psp_ctd_disease_pathways = pd.read_csv('../downloads/val_data/CTD_D013494_pathways_20250904052209.csv') \n",
    "psp_ctd_disease_pathways['Score'] = psp_ctd_disease_pathways['Association inferred via'].apply(lambda x: len(x.split('|'))) \n",
    "psp_ctd_disease_pathways['Entity'] = psp_ctd_disease_pathways.Pathway \n",
    "psp_ctd_disease_pathways = psp_ctd_disease_pathways[['Entity', 'Score']] \n",
    "psp_ctd_disease_pathways.sort_values('Score',  ascending = False, inplace = True) \n",
    "psp_ctd_disease_pathways.insert(0, 'Disease', ['PSP']*psp_ctd_disease_pathways.shape[0])\n",
    "psp_ctd_disease_pathways['Entity'] = psp_ctd_disease_pathways['Entity'].str.strip()\n",
    "psp_ctd_disease_pathways.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "17837f23-d3c4-4963-8a1e-103491d2006f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine and save\n",
    "ctd_disease_pathways = pd.concat([ad_ctd_disease_pathways, brc_ctd_disease_pathways, psp_ctd_disease_pathways], axis = 0)\n",
    "ctd_disease_pathways.to_csv('../data/val_data/CTD_val_data_pathways.csv', index = False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fb475938-655f-4072-9ade-74d50e4058c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD</td>\n",
       "      <td>APP</td>\n",
       "      <td>232.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD</td>\n",
       "      <td>IL1B</td>\n",
       "      <td>148.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AD</td>\n",
       "      <td>CASP3</td>\n",
       "      <td>143.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AD</td>\n",
       "      <td>TNF</td>\n",
       "      <td>142.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AD</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>139.66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Disease Entity   Score\n",
       "0      AD    APP  232.85\n",
       "1      AD   IL1B  148.68\n",
       "2      AD  CASP3  143.60\n",
       "3      AD    TNF  142.44\n",
       "4      AD   BDNF  139.66"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load alzheimer val data (Genes)\n",
    "ctd_alzheimer_genes = pd.read_csv('../downloads/val_data/CTD_D000544_genes_20250721065005.csv') \n",
    "ctd_alzheimer_genes = ctd_alzheimer_genes[['Disease Name', 'Gene Symbol', 'Inference Score']]\n",
    "ctd_alzheimer_genes.rename(columns = {'Disease Name': 'Disease','Gene Symbol': 'Entity', 'Inference Score':'Score' }, inplace = True)\n",
    "ctd_alzheimer_genes['Disease'] = ctd_alzheimer_genes['Disease'].apply(lambda x: x.split(' ')[0])\n",
    "ctd_alzheimer_genes['Disease'] = [\"AD\"]*ctd_alzheimer_genes.shape[0]\n",
    "ctd_alzheimer_genes['Entity'] = ctd_alzheimer_genes['Entity'].str.strip() \n",
    "ctd_alzheimer_genes.dropna(inplace = True)\n",
    "ctd_alzheimer_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5aab13f5-4721-413b-a1d6-27e294d0364a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>TP53</td>\n",
       "      <td>362.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>CCND1</td>\n",
       "      <td>339.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>327.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>BIRC5</td>\n",
       "      <td>308.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>304.61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Disease Entity   Score\n",
       "0    BRCA   TP53  362.21\n",
       "1    BRCA  CCND1  339.15\n",
       "2    BRCA   BCL2  327.23\n",
       "3    BRCA  BIRC5  308.22\n",
       "4    BRCA  PARP1  304.61"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load breast neoplasms val data (Genes)\n",
    "ctd_brca_genes = pd.read_csv('../downloads/val_data/CTD_D001941_genes_20250904053237.csv') \n",
    "ctd_brca_genes = ctd_brca_genes[['Disease Name', 'Gene Symbol', 'Inference Score']]\n",
    "ctd_brca_genes.rename(columns = {'Disease Name': 'Disease','Gene Symbol': 'Entity', 'Inference Score':'Score' }, inplace = True)\n",
    "ctd_brca_genes['Entity'] = ctd_brca_genes['Entity'].str.strip() \n",
    "ctd_brca_genes = ctd_brca_genes[ctd_brca_genes.Disease == \"Breast Neoplasms\"]\n",
    "ctd_brca_genes.Disease = [\"BRCA\"]*ctd_brca_genes.shape[0]\n",
    "ctd_brca_genes.dropna(inplace = True)\n",
    "ctd_brca_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9e240b91-31bb-402b-9342-b3cba2209a96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSP</td>\n",
       "      <td>EIF2AK3</td>\n",
       "      <td>3.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PSP</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>10.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PSP</td>\n",
       "      <td>HEXA</td>\n",
       "      <td>9.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PSP</td>\n",
       "      <td>OTC</td>\n",
       "      <td>9.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PSP</td>\n",
       "      <td>DRD1</td>\n",
       "      <td>8.80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Disease   Entity  Score\n",
       "0      PSP  EIF2AK3   3.12\n",
       "7      PSP    GSK3B  10.22\n",
       "8      PSP     HEXA   9.49\n",
       "9      PSP      OTC   9.08\n",
       "10     PSP     DRD1   8.80"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load Progressive Suprenatal Palsy val data (Genes)\n",
    "ctd_psp_genes = pd.read_csv('../downloads/val_data/CTD_D013494_genes_20250904052250.csv') \n",
    "ctd_psp_genes = ctd_psp_genes[['Disease Name', 'Gene Symbol', 'Inference Score']]\n",
    "ctd_psp_genes.rename(columns = {'Disease Name': 'Disease','Gene Symbol': 'Entity', 'Inference Score':'Score' }, inplace = True)\n",
    "ctd_psp_genes['Entity'] = ctd_psp_genes['Entity'].str.strip() \n",
    "ctd_psp_genes = ctd_psp_genes[ctd_psp_genes.Disease == \"Supranuclear Palsy, Progressive\"]\n",
    "ctd_psp_genes.dropna(inplace = True)\n",
    "ctd_psp_genes.Disease = [\"PSP\"]*ctd_psp_genes.shape[0]\n",
    "ctd_psp_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d6e10461-7e6b-4be2-b236-923d028a2bd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine and save\n",
    "ctd_genes = pd.concat([ctd_alzheimer_genes, ctd_brca_genes, ctd_psp_genes], axis = 0)\n",
    "ctd_genes.to_csv('../data/val_data/CTD_val_data_genes.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5263d0eb-1b5e-4e23-b0d7-311130e30ebd",
   "metadata": {},
   "source": [
    "#### 3. Prepare HMDD Valdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a07793c0-b284-4a94-b5c8-45b8dd8a8fd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1333</th>\n",
       "      <td>Breast Neoplasms</td>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>Colorectal Neoplasms</td>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7624</th>\n",
       "      <td>Carcinoma, Hepatocellular</td>\n",
       "      <td>hsa-mir-122</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2626</th>\n",
       "      <td>Breast Neoplasms</td>\n",
       "      <td>hsa-mir-155</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1366</th>\n",
       "      <td>Carcinoma, Non-Small-Cell Lung</td>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Disease       Entity  Score\n",
       "1333                Breast Neoplasms   hsa-mir-21    121\n",
       "1390            Colorectal Neoplasms   hsa-mir-21     77\n",
       "7624       Carcinoma, Hepatocellular  hsa-mir-122     69\n",
       "2626                Breast Neoplasms  hsa-mir-155     62\n",
       "1366  Carcinoma, Non-Small-Cell Lung   hsa-mir-21     50"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmdd = pd.read_excel('../downloads/val_data/alldata_v4.xlsx') \n",
    "hmdd = hmdd[['disease', 'miRNA']] \n",
    "hmdd['association_count'] = hmdd.groupby(['disease', 'miRNA'])['miRNA'].transform('count')\n",
    "hmdd = hmdd.drop_duplicates().sort_values('association_count', ascending = False) \n",
    "hmdd.rename(columns = {'disease':'Disease', 'miRNA':'Entity', 'association_count': 'Score'}, inplace = True)\n",
    "hmdd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ba27120f-55c6-4c14-bec5-71528631bc84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3881</th>\n",
       "      <td>AD</td>\n",
       "      <td>hsa-mir-146a</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10958</th>\n",
       "      <td>AD</td>\n",
       "      <td>hsa-mir-34a</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11341</th>\n",
       "      <td>AD</td>\n",
       "      <td>hsa-mir-132</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2614</th>\n",
       "      <td>AD</td>\n",
       "      <td>hsa-mir-155</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17541</th>\n",
       "      <td>AD</td>\n",
       "      <td>hsa-mir-107</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Disease        Entity  Score\n",
       "3881       AD  hsa-mir-146a     28\n",
       "10958      AD   hsa-mir-34a     16\n",
       "11341      AD   hsa-mir-132     16\n",
       "2614       AD   hsa-mir-155     12\n",
       "17541      AD   hsa-mir-107     12"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adhmdd = hmdd[hmdd.Disease.isin(['Alzheimer Disease'])] \n",
    "adhmdd['Disease'] = ['AD']*adhmdd.shape[0]\n",
    "brhmdd = hmdd[hmdd.Disease.str.contains('Breast')] \n",
    "brhmdd['Disease'] = ['BRCA']*brhmdd.shape[0]\n",
    "psphmdd = hmdd[hmdd.Disease.isin([\"Supranuclear Palsy, Progressive\"])] \n",
    "psphmdd['Disease'] = ['PSP']*psphmdd.shape[0] \n",
    "\n",
    "hmdd_filtered = pd.concat([adhmdd, brhmdd, psphmdd], axis = 0)   \n",
    "hmdd_filtered['Entity'] = hmdd_filtered['Entity'].str.strip()\n",
    "hmdd_filtered.to_csv('../data/val_data/HMDD_val_data_miRNA.csv', index = False)\n",
    "hmdd_filtered.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2d41569-de36-426e-bc0e-777c982de8e5",
   "metadata": {},
   "source": [
    "#### 4. Prepare EWAS ATLAS Valdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5463bad5-3c12-44dc-b8fb-ba18d672cabd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ewas = pd.read_csv('../downloads/val_data/EWAS_Atlas_associations.tsv', sep='\\t', encoding='ISO-8859-1') \n",
    "ewas = ewas[['trait', 'probe_ID', 'p_value']] \n",
    "ewas.rename(columns = {'trait': 'Disease', 'probe_ID': 'Entity', 'p_value': 'Score'}, inplace = True) \n",
    "ewas.sort_values('Score', ascending = True, inplace =  True)\n",
    "ewas = ewas[ewas.Score >=0].drop_duplicates().dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6067a0dc-ce00-4bc0-a22d-afbf566ab641",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_ewas = ewas[ewas.Disease.str.contains('Alzheimer')]\n",
    "ad_ewas['Disease'] = ['AD']*ad_ewas.shape[0]\n",
    " \n",
    "breast_cancer_terms = ['breast cancer', 'breast cancer overall survival', \n",
    "       'chemotherapy for breast cancer', 'breast cancerization',\n",
    "       'non-invasive sporadic breast cancer', 'Breast Cancer',\n",
    "       'invasive lobular breast cancer overall survival',\n",
    "       'aerobic training for breast cancer', 'breast tumours']\n",
    "brc_ewas = ewas[ewas.Disease.isin(breast_cancer_terms)] \n",
    "brc_ewas['Disease'] = ['BRCA']*brc_ewas.shape[0]\n",
    "\n",
    " \n",
    "psp_ewas = ewas[ewas.Disease.str.contains(\"progressive supranuclear palsy\")] \n",
    "psp_ewas['Disease'] = ['PSP']*psp_ewas.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2ed26c3a-40a3-48c1-9855-57f4750252e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease</th>\n",
       "      <th>Entity</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>780686</th>\n",
       "      <td>AD</td>\n",
       "      <td>cg23971107</td>\n",
       "      <td>4.090000e-97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780683</th>\n",
       "      <td>AD</td>\n",
       "      <td>cg16100476</td>\n",
       "      <td>4.200000e-97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780685</th>\n",
       "      <td>AD</td>\n",
       "      <td>cg17258228</td>\n",
       "      <td>4.230000e-97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761125</th>\n",
       "      <td>AD</td>\n",
       "      <td>cg04399985</td>\n",
       "      <td>4.920000e-97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780680</th>\n",
       "      <td>AD</td>\n",
       "      <td>cg01437515</td>\n",
       "      <td>5.170000e-97</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Disease      Entity         Score\n",
       "780686      AD  cg23971107  4.090000e-97\n",
       "780683      AD  cg16100476  4.200000e-97\n",
       "780685      AD  cg17258228  4.230000e-97\n",
       "761125      AD  cg04399985  4.920000e-97\n",
       "780680      AD  cg01437515  5.170000e-97"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# concat ewas data\n",
    "ewas_filtered = pd.concat([ad_ewas, brc_ewas, psp_ewas], axis =0)   \n",
    "ewas_filtered['Entity'] = ewas_filtered['Entity'].str.strip()\n",
    "ewas_filtered.to_csv('../data/val_data/EWAS_ATLAS_va_data.csv', index = False)\n",
    "ewas_filtered.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
