{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c9ee4a9a-e7f4-4263-bb04-676342b8dd22",
   "metadata": {},
   "source": [
    "### Import Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3683f39f-bdce-4ff4-86c9-d942ab31a145",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import warnings\n",
    "from modules.utils import save_json\n",
    "warnings.filterwarnings('ignore') \n",
    "\n",
    "\n",
    "import numpy as np\n",
    "def downsample_majority_to_ratio(\n",
    "    df: pd.DataFrame,\n",
    "    label_col: str = \"Diagnosis\",\n",
    "    control_label: str = \"Control\",\n",
    "    tumor_label: str = \"Tumor\",   # use \"AD\" here if that's your positive label\n",
    "    ratio: float = 1.10,\n",
    "    random_state: int = 42\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"Keep all Control; downsample Tumor to â‰ˆ ratio * n_control.\"\"\"\n",
    "    df_ctrl = df[df[label_col] == control_label]\n",
    "    df_tum  = df[df[label_col] == tumor_label]\n",
    "\n",
    "    n_ctrl = len(df_ctrl)\n",
    "    target_tum = int(np.ceil(ratio * n_ctrl))\n",
    "\n",
    "    # If we have more tumor than target, sample down; otherwise keep as is\n",
    "    if len(df_tum) > target_tum:\n",
    "        df_tum = df_tum.sample(n=target_tum, random_state=random_state)\n",
    "\n",
    "    out = pd.concat([df_ctrl, df_tum], axis=0).sample(frac=1, random_state=random_state).reset_index(drop=True)\n",
    "    return out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db6caf27-ccea-48eb-aeaf-6f03a89b0d71",
   "metadata": {},
   "source": [
    "#### 1. Prepare ROSMAP DATASETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f8be5a12-eec9-4a81-9721-888338c57259",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load rosmap datasets\n",
    "miRNA = pd.read_csv('../data/ROSMAP/preprocessed/miRNA_data.csv').drop(columns = ['Gender',\t'Race',\t'PMI',\t'Braak'])\n",
    "mRNA = pd.read_csv('../data/ROSMAP/preprocessed/gene_expression_data.csv').drop(columns = ['Gender',\t'Race',\t'PMI',\t'Braak'])\n",
    "meth = pd.read_csv('../data/ROSMAP/preprocessed/dna_methylation_data.csv').drop(columns = ['Gender',\t'Race',\t'PMI',\t'Braak'])\n",
    "\n",
    "# a few samples have repeats in mRNA\n",
    "df = mRNA.copy()\n",
    "\n",
    "# Separate categorical and numeric columns\n",
    "categorical_cols = ['Diagnosis']  # adjust if needed\n",
    "numeric_cols = df.columns.difference(['Sample ID'] + categorical_cols)\n",
    "\n",
    "# Aggregate numeric columns by mean\n",
    "df_numeric = df.groupby(\"Sample ID\", as_index=False)[numeric_cols].mean()\n",
    "\n",
    "# For categorical columns, take the first occurrence for each Sample ID\n",
    "df_categorical = df.groupby(\"Sample ID\", as_index=False)[categorical_cols].first()\n",
    "\n",
    "# Merge numeric and categorical data back together\n",
    "mRNA = pd.merge(df_categorical, df_numeric, on=\"Sample ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a6d47fd8-e885-4452-b574-0bab4f4fc616",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of miRNA samples: 378\n",
      "Number of gene expression samples: 378\n",
      "Number of dna methylation samples: 375\n",
      "Number of miRNA and gene expression Multiltomics samples: 378\n",
      "Number of miRNA and dna methylation Multiltomics samples: 378\n",
      "Number of gene expression and dna methylation Multiltomics samples: 375\n",
      "Number of miRNA, gene expression and dna methylation Multiltomics samples: 375\n"
     ]
    }
   ],
   "source": [
    "# save experimental data\n",
    "\n",
    "# save single omics\n",
    "miRNA.to_csv('../data/ROSMAP/prepared/miRNA_data.csv', index=False)\n",
    "mRNA.to_csv('../data/ROSMAP/prepared/gene_expression_data.csv', index=False)\n",
    "meth.to_csv('../data/ROSMAP/prepared/dna_methylation_data.csv', index=False)  \n",
    "\n",
    "# select and save samples\n",
    "miRNA_and_gene_expression_samples = list(set(miRNA['Sample ID']) & set(mRNA['Sample ID']))\n",
    "miRNA_and_dna_methylation_samples = list(set(miRNA['Sample ID']) & set(meth['Sample ID']))\n",
    "gene_expression_and_dna_methylation_samples = list(set(mRNA['Sample ID']) & set(meth['Sample ID'])) \n",
    "miRNA_gene_expression_and_dna_methylation_samples = list(set(miRNA['Sample ID']) & set(mRNA['Sample ID']) & set(meth['Sample ID'])) \n",
    "\n",
    "\n",
    "pd.merge(miRNA, mRNA).to_csv('../data/ROSMAP/prepared/miRNA_and_gene_expression_data.csv', index=False)\n",
    "pd.merge(miRNA, meth).to_csv('../data/ROSMAP/prepared/miRNA_and_dna_methylation_data.csv', index=False)\n",
    "pd.merge(mRNA, meth).to_csv('../data/ROSMAP/prepared/gene_expression_and_dna_methylation_data.csv', index=False)\n",
    "pd.merge(miRNA, pd.merge(mRNA, meth)).to_csv('../data/ROSMAP/prepared/miRNA_and_gene_expression_and_dna_methylation_data.csv', index=False)\n",
    "\n",
    "print(f'Number of miRNA samples: {miRNA.shape[0]}')\n",
    "print(f'Number of gene expression samples: {mRNA.shape[0]}')\n",
    "print(f'Number of dna methylation samples: {meth.shape[0]}')\n",
    "\n",
    "print(f'Number of miRNA and gene expression Multiltomics samples: {len(miRNA_and_gene_expression_samples)}')\n",
    "print(f'Number of miRNA and dna methylation Multiltomics samples: {len(miRNA_and_gene_expression_samples)}')\n",
    "print(f'Number of gene expression and dna methylation Multiltomics samples: {len(gene_expression_and_dna_methylation_samples )}')\n",
    "print(f'Number of miRNA, gene expression and dna methylation Multiltomics samples: {len(miRNA_gene_expression_and_dna_methylation_samples)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c174fc8d-d3aa-40ff-a37b-ebd597915812",
   "metadata": {},
   "source": [
    "#### 2. Prepare MayoRNASeq DATASETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b2051642-80ed-48b3-80c1-c215e5f6f3de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load rosmap datasets\n",
    "metabolomics = pd.read_csv('../data/MayoRNASeq/preprocessed/metabolomics_data.csv').drop(columns = ['Gender',\t'ageDeath', 'Braak'])\n",
    "mRNA = pd.read_csv('../data/MayoRNASeq/preprocessed/gene_expression_data.csv').drop(columns = ['Gender',\t'ageDeath', 'Braak'])\n",
    "proteomics = pd.read_csv('../data/MayoRNASeq/preprocessed/proteomics_data.csv').drop(columns = ['Gender',\t'ageDeath', 'Braak']) \n",
    "metabolomics['Sample ID'] = metabolomics['Sample ID'].astype(str)\n",
    "mRNA['Sample ID'] = mRNA['Sample ID'].astype(str)\n",
    "proteomics['Sample ID'] = proteomics['Sample ID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b1d591d8-e94f-47c6-92c9-df00c2eaa953",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of metabolomics samples: 98\n",
      "Number of gene expression samples: 162\n",
      "Number of proteomics samples: 112\n",
      "Number of metabolomics and gene expression Multiltomics samples: 98\n",
      "Number of metabolomics and proteomics Multiltomics samples: 98\n",
      "Number of gene expression and proteomics Multiltomics samples: 112\n",
      "Number of metabolomics, gene expression and proteomics Multiltomics samples: 97\n"
     ]
    }
   ],
   "source": [
    "# save experimental data\n",
    "\n",
    "# save single omics\n",
    "metabolomics.to_csv('../data/MayoRNASeq/prepared/metabolomics_data.csv', index=False)\n",
    "mRNA.to_csv('../data/MayoRNASeq/prepared/gene_expression_data.csv', index=False)\n",
    "proteomics.to_csv('../data/MayoRNASeq/prepared/proteomics_data.csv', index=False)  \n",
    "\n",
    "# select and save samples\n",
    "metabolomics_and_gene_expression_samples = list(set(metabolomics['Sample ID']) & set(mRNA['Sample ID']))\n",
    "metabolomics_and_proteomics_samples = list(set(metabolomics['Sample ID']) & set(proteomics['Sample ID']))\n",
    "gene_expression_and_proteomics_samples = list(set(mRNA['Sample ID']) & set(proteomics['Sample ID'])) \n",
    "metabolomics_gene_expression_and_proteomics_samples = list(set(metabolomics['Sample ID']) & set(mRNA['Sample ID']) & set(proteomics['Sample ID'])) \n",
    "\n",
    "\n",
    "df1 = pd.merge(metabolomics, mRNA, on='Sample ID')\n",
    "df1.rename(columns={'Diagnosis_x':'Diagnosis'}, inplace=True)\n",
    "df1.drop(columns=['Diagnosis_y'], inplace=True)\n",
    "df1.to_csv('../data/MayoRNASeq/prepared/metabolomics_and_gene_expression_data.csv', index=False)\n",
    "\n",
    "df2 = pd.merge(metabolomics, proteomics, on='Sample ID') \n",
    "df2.rename(columns={ column: column.split('_x')[0] for column in df2.columns if '_x' in column}, inplace=True) \n",
    "df2.drop(columns=['Diagnosis_y'], inplace=True)\n",
    "df2.to_csv('../data/MayoRNASeq/prepared/metabolomics_and_proteomics_data.csv', index=False)\n",
    "\n",
    "df3= pd.merge(mRNA, proteomics, on='Sample ID') \n",
    "df3.rename(columns={'Diagnosis_x':'Diagnosis'}, inplace=True) \n",
    "df3.drop(columns=['Diagnosis_y'], inplace=True)\n",
    "df3.to_csv('../data/MayoRNASeq/prepared/gene_expression_and_proteomics_data.csv', index=False)\n",
    "\n",
    "df4 = pd.merge(metabolomics, pd.merge(mRNA, proteomics, on='Sample ID'), on='Sample ID')   \n",
    "df4.drop(columns=['Diagnosis_y', 'Diagnosis_x'], inplace=True) \n",
    "df4.rename(columns={ column: column.split('_x')[0] for column in df4.columns if '_x' in column}, inplace=True)\n",
    "df4.to_csv('../data/MayoRNASeq/prepared/metabolomics_and_gene_expression_and_proteomics_data.csv', index=False)\n",
    "\n",
    "print(f'Number of metabolomics samples: {metabolomics.shape[0]}')\n",
    "print(f'Number of gene expression samples: {mRNA.shape[0]}')\n",
    "print(f'Number of proteomics samples: {proteomics.shape[0]}')\n",
    "\n",
    "print(f'Number of metabolomics and gene expression Multiltomics samples: {len(metabolomics_and_gene_expression_samples)}')\n",
    "print(f'Number of metabolomics and proteomics Multiltomics samples: {len(metabolomics_and_gene_expression_samples)}')\n",
    "print(f'Number of gene expression and proteomics Multiltomics samples: {len(gene_expression_and_proteomics_samples )}')\n",
    "print(f'Number of metabolomics, gene expression and proteomics Multiltomics samples: {len(metabolomics_gene_expression_and_proteomics_samples)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f13a4e17-04f0-4c49-a73a-881e06f9d615",
   "metadata": {},
   "source": [
    "#### 3. Prepare BRCA DATASETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7659ba26-1cbf-45cb-babf-18af2830a875",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of miRNA samples before downsampling: 819\n",
      "Number of gene expression samples before downsampling: 1202\n",
      "Number of dna methylation samples before downsampling: 877\n"
     ]
    }
   ],
   "source": [
    "# load rosmap datasets\n",
    "miRNA = pd.read_csv('../data/BRCA/preprocessed/miRNA_expression_data.csv').drop(columns = ['Age','Gender','Stage'])\n",
    "mRNA = pd.read_csv('../data/BRCA/preprocessed/gene_expression_data.csv').drop(columns = ['Age','Gender','Stage'])\n",
    "meth = pd.read_csv('../data/BRCA/preprocessed/dna_methylation_data.csv').drop(columns = ['Age','Gender','Stage'])\n",
    "\n",
    "print(f'Number of miRNA samples before downsampling: {miRNA.shape[0]}')\n",
    "print(f'Number of gene expression samples before downsampling: {mRNA.shape[0]}')\n",
    "print(f'Number of dna methylation samples before downsampling: {meth.shape[0]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "da56bdb9-cea9-459f-a299-5b456a849e55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# balance dataset\n",
    "miRNA1 = downsample_majority_to_ratio(\n",
    "    df=miRNA,\n",
    "    label_col=\"Diagnosis\",\n",
    "    control_label=\"Control\",\n",
    "    tumor_label=\"Tumor\",  \n",
    "    ratio=1.10,\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# balance dataset\n",
    "mRNA1 = downsample_majority_to_ratio(\n",
    "    df=mRNA,\n",
    "    label_col=\"Diagnosis\",\n",
    "    control_label=\"Control\",\n",
    "    tumor_label=\"Tumor\",  \n",
    "    ratio=1.10,\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# balance dataset\n",
    "meth1 = downsample_majority_to_ratio(\n",
    "    df=meth,\n",
    "    label_col=\"Diagnosis\",\n",
    "    control_label=\"Control\",\n",
    "    tumor_label=\"Tumor\",  \n",
    "    ratio=1.10,\n",
    "    random_state=42\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3222348e-afb9-48fd-806f-12bd4a75eb21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Diagnosis\n",
       "Tumor      57\n",
       "Control    51\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df4.Diagnosis.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3fb23d50-4545-46d9-b1ef-faf0af54e264",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of miRNA samples: 158\n",
      "Number of gene expression samples: 236\n",
      "Number of dna methylation samples: 202\n",
      "Number of miRNA and gene expression Multiltomics samples: 158\n",
      "Number of miRNA and dna methylation Multiltomics samples: 108\n",
      "Number of gene expression and dna methylation Multiltomics samples: 175\n",
      "Number of miRNA, gene expression and dna methylation Multiltomics samples: 108\n"
     ]
    }
   ],
   "source": [
    "# save experimental data\n",
    "\n",
    "# save single omics\n",
    "miRNA1.to_csv('../data/BRCA/prepared/miRNA_data.csv', index=False)\n",
    "mRNA1.to_csv('../data/BRCA/prepared/gene_expression_data.csv', index=False)\n",
    "meth1.to_csv('../data/BRCA/prepared/dna_methylation_data.csv', index=False)  \n",
    "\n",
    "# select and save samples\n",
    "miRNA_and_gene_expression_samples = list(set(miRNA['Sample ID']) & set(mRNA['Sample ID']))\n",
    "miRNA_and_dna_methylation_samples = list(set(miRNA['Sample ID']) & set(meth['Sample ID']))\n",
    "gene_expression_and_dna_methylation_samples = list(set(mRNA['Sample ID']) & set(meth['Sample ID'])) \n",
    "miRNA_gene_expression_and_dna_methylation_samples = list(set(miRNA['Sample ID']) & set(mRNA['Sample ID']) & set(meth['Sample ID'])) \n",
    "\n",
    "df1 = pd.merge(miRNA, mRNA)\n",
    "df1 = downsample_majority_to_ratio(\n",
    "    df=df1,\n",
    "    label_col=\"Diagnosis\",\n",
    "    control_label=\"Control\",\n",
    "    tumor_label=\"Tumor\",  \n",
    "    ratio=1.10,\n",
    "    random_state=42\n",
    ")\n",
    "df1.to_csv('../data/BRCA/prepared/miRNA_and_gene_expression_data.csv', index=False)\n",
    "\n",
    "df2 = pd.merge(miRNA, meth)\n",
    "df2 = downsample_majority_to_ratio(\n",
    "    df=df2,\n",
    "    label_col=\"Diagnosis\",\n",
    "    control_label=\"Control\",\n",
    "    tumor_label=\"Tumor\",  \n",
    "    ratio=1.10,\n",
    "    random_state=42\n",
    ")\n",
    "df2.to_csv('../data/BRCA/prepared/miRNA_and_dna_methylation_data.csv', index=False)\n",
    "\n",
    "df3 = pd.merge(mRNA, meth)\n",
    "df3 = downsample_majority_to_ratio(\n",
    "    df=df3,\n",
    "    label_col=\"Diagnosis\",\n",
    "    control_label=\"Control\",\n",
    "    tumor_label=\"Tumor\",  \n",
    "    ratio=1.10,\n",
    "    random_state=42\n",
    ")\n",
    "df3.to_csv('../data/BRCA/prepared/gene_expression_and_dna_methylation_data.csv', index=False) \n",
    "\n",
    "df4 = pd.merge(miRNA, pd.merge(mRNA, meth))\n",
    "df4 = downsample_majority_to_ratio(\n",
    "    df=df4,\n",
    "    label_col=\"Diagnosis\",\n",
    "    control_label=\"Control\",\n",
    "    tumor_label=\"Tumor\",  \n",
    "    ratio=1.10,\n",
    "    random_state=42\n",
    ")\n",
    "df4.to_csv('../data/BRCA/prepared/miRNA_and_gene_expression_and_dna_methylation_data.csv', index=False)\n",
    "\n",
    "print(f'Number of miRNA samples: {miRNA1.shape[0]}')\n",
    "print(f'Number of gene expression samples: {mRNA1.shape[0]}')\n",
    "print(f'Number of dna methylation samples: {meth1.shape[0]}')\n",
    "\n",
    "print(f'Number of miRNA and gene expression Multiltomics samples: {df1.shape[0]}')\n",
    "print(f'Number of miRNA and dna methylation Multiltomics samples: {df2.shape[0]}')\n",
    "print(f'Number of gene expression and dna methylation Multiltomics samples: {df3.shape[0]}')\n",
    "print(f'Number of miRNA, gene expression and dna methylation Multiltomics samples: {df4.shape[0]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "30a50d4f-ad21-4311-91fb-29e4a242c6d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## feature names\n",
    "feature_names = {\n",
    "    1:df1.columns[1:].to_list(),\n",
    "    2:df2.columns[1:].to_list(),\n",
    "    3:df3.columns[1:].to_list()\n",
    "}\n",
    "save_json(\"../results/BRCA/featurenames.json\", feature_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5294c4b5-8c90-4003-83db-d5c7f0ab9d6c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
